### PR TITLE
ont hot support tensor depth

### PR DESCRIPTION
**Pull request by phlrain**
ont hot op的depth 参数支持tensor 类型

layers.one_hot( a, depth=depth)

depth 可以支持tensor的输入

### CODE REVIEW

File: paddle/fluid/framework/framework.proto [hunk 1]
```diff
@@ -82,6 +82,9 @@ message OpProto {
     optional bool duplicable = 3 [ default = false ];
     optional bool intermediate = 4 [ default = false ];
     optional bool dispensable = 5 [ default = false ];
+    optional bool remain_cpu = 6 [ default = false ];
```
Comment (Superjomn) [inline paddle/fluid/framework/framework.proto]: 非模型算法需要，不建议修改 framework.proto 
静态模型不需要的动态执行信息

File: paddle/fluid/framework/framework.proto [hunk 2]
```diff
@@ -82,6 +82,9 @@ message OpProto {
...
+    optional string type = 7;
+    optional bool only_in_static_model = 8 [ default = false ];
   }
```
Comment (phlrain) [inline paddle/fluid/framework/framework.proto]: proto 修改这个，春伟不建议这么修改，我会改一个版本提交上来

File: paddle/fluid/framework/framework.proto (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/framework.proto]: 赞成 @Superjomn

File: paddle/fluid/framework/operator.cc [hunk 3]
```diff
@@ -1018,6 +1018,19 @@ void OperatorWithKernel::TransferInplaceVarsBack(
   }
 }
 
+bool OperatorWithKernel::is_remain_cpu(const std::string& str_name) const {
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc]: What's the meaning of `remain_cpu`?

File: paddle/fluid/framework/operator.cc [hunk 4]
```diff
@@ -1060,13 +1074,16 @@ Scope* OperatorWithKernel::PrepareData(
...
         continue;
       }
-
       auto kernel_type_for_var = GetKernelTypeForVar(
           var_name_item.first, *tensor_in, expected_kernel_key);
 
       if (!NeedTransform(kernel_type_for_var, expected_kernel_key)) {
         continue;
       }
+      if (is_remain_cpu((const std::string)var_name_item.first) == true &&
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc]: Why do you use `(const std::string)` here?

File: paddle/fluid/operators/one_hot_op.cc [hunk 5]
```diff
@@ -52,11 +63,22 @@ class OneHotOpMaker : public framework::OpProtoAndCheckerMaker {
...
-    AddAttr<int>("depth",
-                 "A positive integer to specify the length of one-hot vector.");
+
+    AddAttr<int>("depth_attr",
+                 "A positive integer to specify the length of one-hot vector.")
+        .SetDefault(-1);
+    AddAttr<bool>(
+        "use_attr",
+        "An boolen to special to use depth in attr or use depth in Input")
+        .SetDefault(false);
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cc]: Why do you change the attributes and input? you should consider the compatibility.

File: python/paddle/fluid/framework.py [hunk 6]
```diff
@@ -1027,7 +1071,7 @@ def find_name(var_list, name):
                     if not ((m.name in outputs) or m.dispensable):
                         raise ValueError(("Incorrect setting for output(s) of "
                                           "operator \"%s\", should set: [%s].")
-                                         % (type, m.name))
+                                         % (op_type, m.name))
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py]: Why change to `op_type` here?

File: python/paddle/fluid/framework.py [hunk 7]
```diff
@@ -1009,13 +1019,47 @@ def find_name(var_list, name):
...
+
+                                in_arg_names.append(cpt.to_text(out_var.name))
+
+                            else:
+
+                                raise ValueError(
+                                    "not suprt args type , should be[ string_type, binary_type, Varibale] type is"
+                                )
+                            #elif in_proto.type is not None and in_proto.type != "" and \
+                            #        isinstance( arg)
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py]: If line 1061 and line 1062 are unnecessary, they should be removed.

File: python/paddle/fluid/framework.py [hunk 8]
```diff
@@ -1009,13 +1019,47 @@ def find_name(var_list, name):
...
+                                    desc=fill_op_desc,
+                                    type='fill_constant',
+                                    inputs={},
+                                    outputs={'Out': [out_var]},
+                                    attrs={
+                                        'shape': [1],
+                                        'dtype': out_var.dtype,
+                                        'value': float(arg),
+                                        'force_cpu': True,
+                                    })
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py]: Why? The code looks a little trick.

File: paddle/fluid/operators/one_hot_op.cc [hunk 9]
```diff
@@ -52,11 +68,18 @@ class OneHotOpMaker : public framework::OpProtoAndCheckerMaker {
...
               "The tensor consists of one-hot representations of values in X.");
-    AddAttr<int>("depth",
-                 "A positive integer to specify the length of one-hot vector.");
+
+    AddAttr<int>("depth_attr",
+                 "A positive integer to specify the length of one-hot vector.")
+        .SetDefault(-1);
+    AddAttr<std::vector<std::string>>("remain_cpu_name_list",
+                                      "var name list remain cpu")
+        .SetDefault(std::vector<std::string>());
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cc]: Do you consider the compatibility problems？

File: python/paddle/fluid/tests/unittests/test_one_hot_op.py [hunk 10]
```diff
@@ -96,7 +98,8 @@ def test_check_output(self):
             block.append_op(
                 type='one_hot',
                 inputs={'X': x},
-                attrs={'depth': self.depth},
+                attrs={},
+                inputs_or_attr={'depth': self.depth},
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_one_hot_op.py]: What's the meaning of `inputs_or_attr `?

File: paddle/fluid/framework/operator.cc [hunk 11]
```diff
@@ -1055,6 +1055,7 @@ Scope* OperatorWithKernel::PrepareData(
       auto* var = input_vars[i];
 
       // Only tensor can be tranfer to another device.
+
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc]: Please don't change this file.
```
git checkout develop  paddle/fluid/framework/operator.cc 
```

File: paddle/fluid/operators/one_hot_op.cc [hunk 12]
```diff
@@ -24,6 +26,8 @@ class OneHotOp : public framework::OperatorWithKernel {
   void InferShape(framework::InferShapeContext* ctx) const override {
     PADDLE_ENFORCE(ctx->HasInput("X"),
                    "Input(X) of OneHotOp should not be null.");
+    // PADDLE_ENFORCE(ctx->HasInput("depth"),
+    //               "Input(depth) of depth should not be null.");
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cc]: If line29 and line 30 are unnecessary, please remove them.

File: paddle/fluid/operators/one_hot_op.cc [hunk 13]
```diff
@@ -34,15 +38,35 @@ class OneHotOp : public framework::OperatorWithKernel {
       PADDLE_ENFORCE_GE(x_dims[x_dims.size() - 1], 1U,
                         "Last dimension of Input(X) should be 1.");
     }
-    int depth = ctx->Attrs().Get<int>("depth");
-
-    PADDLE_ENFORCE_GT(depth, 0, "Should provide a positive depth (%d).", depth);
 
     framework::DDim out_dims(x_dims);
+    int depth = -1;
+    if (!ctx->HasInput("depth_tensor")) {
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cc]: I prefer to rename `depth_tensor` with `depth`.

File: paddle/fluid/operators/one_hot_op.cc [hunk 14]
```diff
@@ -34,15 +38,35 @@ class OneHotOp : public framework::OperatorWithKernel {
...
+
+  framework::OpKernelType GetKernelTypeForVar(
+      const std::string& var_name, const Tensor& tensor,
+      const framework::OpKernelType& expected_kernel_type) const override {
+    if (var_name == "depth_tensor") {
+      return expected_kernel_type;
+    } else {
+      return framework::OpKernelType(expected_kernel_type.data_type_,
+                                     tensor.place(), tensor.layout());
+    }
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cc:63]:
```
    if (var_name == "depth_tensor") {
      return expected_kernel_type;
    } 
    return framework::OpKernelType(expected_kernel_type.data_type_,
                                     tensor.place(), tensor.layout());

```

File: paddle/fluid/operators/one_hot_op.cu [hunk 15]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
   void Compute(const framework::ExecutionContext& context) const override {
     auto* in = context.Input<LoDTensor>("X");
     auto* out = context.Output<LoDTensor>("Out");
-    int depth = context.Attr<int>("depth");
+    // int depth = context.Attr<int>("depth");
 
+    int depth = -1;
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu:66]: line67 should be removed.
```
int depth = context.Attr<int>("depth");
```

File: paddle/fluid/operators/one_hot_op.cu [hunk 16]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
...
+            platform::DeviceContextPool::Instance();
+        auto place = platform::CUDAPlace();
+        auto& dev_ctx = *pool.Get(place);
+        framework::TensorCopy(*depth_tensor, platform::CPUPlace(), dev_ctx,
+                              &temp);
+        dev_ctx.Wait();
+
+        depth = static_cast<int32_t>(*temp.data<int32_t>());
+
+        LOG(ERROR) << "in gpu";
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu]: Why use `LOG(ERROR) << "in gpu";` here?

File: paddle/fluid/operators/one_hot_op.cu [hunk 17]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
...
+
+      } else {
+        depth = static_cast<int32_t>(*depth_tensor->data<int32_t>());
+        LOG(ERROR) << "in cpu";
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu]: Why use `LOG(ERROR) << "in cpu";` here?

File: paddle/fluid/operators/one_hot_op.cu [hunk 18]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
...
+    // int depth = context.Attr<int>("depth");
 
+    int depth = -1;
+    if (context.HasInput("depth_tensor")) {
+      auto* depth_tensor = context.Input<framework::Tensor>("depth_tensor");
+      if (platform::is_gpu_place(depth_tensor->place())) {
+        framework::Tensor temp;
+        platform::DeviceContextPool& pool =
+            platform::DeviceContextPool::Instance();
+        auto place = platform::CUDAPlace();
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu]: ->`auto place = platform::CPUPlace();`

File: paddle/fluid/operators/one_hot_op.cu [hunk 19]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
...
+        auto& dev_ctx = *pool.Get(place);
+        framework::TensorCopy(*depth_tensor, platform::CPUPlace(), dev_ctx,
+                              &temp);
+        dev_ctx.Wait();
+
+        depth = static_cast<int32_t>(*temp.data<int32_t>());
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu]: Why the dtype of `temp` must be `int32_t `?

Comment (chengduoZH) [reply paddle/fluid/operators/one_hot_op.cu]: Why use ` static_cast<int32_t>` here?

File: paddle/fluid/operators/one_hot_op.cu [hunk 20]
```diff
@@ -62,8 +62,37 @@ class OneHotCUDAKernel : public framework::OpKernel<T> {
...
+
+        LOG(ERROR) << "in gpu";
+
+      } else {
+        depth = static_cast<int32_t>(*depth_tensor->data<int32_t>());
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.cu]: Why use ` static_cast<int32_t>`?

File: paddle/fluid/operators/one_hot_op.h [hunk 21]
```diff
@@ -49,13 +49,25 @@ struct OneHotOpFunctor {
...
+      auto* depth_tensor = context.Input<Tensor>("depth_tensor");
+      auto* depth_data = depth_tensor->data<int32_t>();
+      depth = depth_data[0];
+      auto in_dims = in->dims();
+      framework::DDim out_dims(in_dims);
+      out_dims[out_dims.size() - 1] = depth;
+      out->Resize(out_dims);
+    } else {
+      depth = context.Attr<int>("depth");
+    }
```
Comment (chengduoZH) [inline paddle/fluid/operators/one_hot_op.h:68]:
```
    int depth =  context.Attr<int>("depth");
    if (context.HasInput("depth_tensor")) {
      auto* depth_tensor = context.Input<Tensor>("depth_tensor");
      auto* depth_data = depth_tensor->data<int32_t>();
      framework::DDim out_dims( in->dims());
      out_dims[out_dims.size() - 1] = depth_data[0];
      out->Resize(out_dims);
    }
```

File: python/paddle/fluid/framework.py [hunk 22]
```diff
@@ -1035,8 +1035,18 @@ def find_name(var_list, name):
...
+                        def convert_type(data_type):
+                            if data_type == "int32" or data_type == 'int64':
+                                return eval('int')
+                            elif data_type == 'float32' or data_type == 'float64':
+                                return eval('float')
+                            else:
+                                raise ValueError(
+                                    "Type {} not support, should be one of [int32, int64, float32, float64]".
+                                    format(data_type))
+
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py]: `convert_type ` is not used in this PR, it should be removed.

File: python/paddle/fluid/layers/nn.py [hunk 23]
```diff
@@ -6355,10 +6355,33 @@ def sampled_softmax_with_cross_entropy(logits,
...
+            attrs={
+                'shape': [1],
+                'dtype': depth_var.dtype,
+                'value': float(num_samples + 1),
+                'force_cpu': True,
+            },
+            stop_gradient=True)
+        depth_var.stop_gradient = True
+
+        one_hot_input['depth_tensor'] = depth_var
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/nn.py]: What's the meaning of line 6365~6379?
Why do you change the origin implementation?

File: python/paddle/fluid/layers/nn.py [hunk 24]
```diff
@@ -6447,11 +6470,40 @@ def one_hot(input, depth):
...
+                inputs={},
+                outputs={'Out': [depth_var]},
+                attrs={
+                    'shape': [1],
+                    'dtype': depth_var.dtype,
+                    'value': float(depth),
+                    'force_cpu': True,
+                },
+                stop_gradient=True)
+            depth_var.stop_gradient = True
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/nn.py]: Same above.

File: python/paddle/fluid/tests/unittests/test_one_hot_op.py [hunk 25]
```diff
@@ -51,19 +52,20 @@ class TestOneHotOp_default_dtype(OpTest):
...
         out = np.zeros(shape=(np.product(x.shape[:-1]),
                               depth)).astype('float32')
 
         for i in range(np.product(x.shape)):
             out[i, x[i]] = 1.0
 
-        self.inputs = {'X': (x, x_lod)}
-        self.attrs = {'depth': depth}
+        self.inputs = {'X': (x, x_lod), 'depth_tensor': depth_np}
+        self.attrs = {}
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_one_hot_op.py:91]: This is not the situation that we should only make the depth as the input but not attribute.

File: python/paddle/fluid/tests/unittests/test_one_hot_op.py [hunk 26]
```diff
@@ -40,21 +64,45 @@ def setUp(self):
             out[i, x[i]] = 1.0
 
         self.inputs = {'X': (x, x_lod)}
-        self.attrs = {'depth': depth, 'dtype': int(core.VarDesc.VarType.FP32)}
+        self.attrs = {'dtype': int(core.VarDesc.VarType.FP32), 'depth': depth}
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_one_hot_op.py:67]: Why do you change the order?

**Review by chengduoZH — APPROVED** [review]
LGTM

**Pull request merged**